### PR TITLE
Implement concatenation algorithm for paths.

### PR DESCRIPTION
**Pull request by patrickt**
I wrote this in a very literate style for clarity's sake.

### CODE REVIEW

File: semantic-scope-graph/src/Data/Functor/Tagged.hs [hunk 1]
```diff
@@ -63,3 +65,7 @@ taggedM a = (a :#) <$> fresh
 -- ordered, but are guaranteed to be unique throughout the life of the program.
 taggedIO :: a -> IO (Tagged a)
 taggedIO a = (a :#) . hashUnique <$> newUnique
+
+-- This is bad. Don't use it.
+unsafeTagged :: a -> Tagged a
+unsafeTagged = unsafePerformIO . taggedIO
```
Comment (patrickt) [inline semantic-scope-graph/src/Data/Functor/Tagged.hs]: Nothing to see here. Move along.

**Comment by dcreager** [comment]
(Note that my :+1: is pretty cursory and high-level — someone else should handle the deeper Haskell code review part)

**Comment by tclem** [comment]
Broad strokes of the algorithm look right, let's add some tests to make sure we've got the various edge cases!
